### PR TITLE
feat: gas estimates

### PR DESCRIPTION
**Pull request by joepegler**
## Gas estimates for userOps

```
const amountInWei = await smartAccount.getGasEstimates([tx, tx], {
  paymasterServiceData: {
    mode: PaymasterMode.SPONSORED,
  },
});
```

**Checklist:**

- [x] I have performed a self-review of my own code.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [X] I have made corresponding changes to the documentation, if applicable.
- [x] My changes generate no new warnings or errors.

---

## PR-Codex overview
The focus of this PR is to enhance error handling in HTTP requests and improve gas estimation for user operations.

### Detailed summary
- Improved error handling in HTTP requests with additional status and service information
- Added a new method for gas estimation in user operations
- Updated Paymaster class with a new mode constant and added a new token constant
- Updated test cases for gas estimation and error handling
- Fixed typos in comments and URLs in the code

> The following files were skipped due to too many changes: `src/account/BiconomySmartAccountV2.ts`

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

File: src/account/BiconomySmartAccountV2.ts [hunk 1]
```diff
@@ -346,6 +346,81 @@ export class BiconomySmartAccountV2 extends BaseSmartContractAccount {
...
+    if (buildUseropDto?.paymasterServiceData?.mode) {
+      return (
+        _callGasLimit +
+        _preVerificationGas +
+        _verificationGasLimit * _maxFeePerGas
+      )
+    }
+    return (
+      (_callGasLimit +
+        BigInt(3) * _verificationGasLimit +
```
Comment (VGabriel45) [inline src/account/BiconomySmartAccountV2.ts:418]: What's the reason behind _verificationGasLimit 3x ? Isn't it too much ?

File: src/account/BiconomySmartAccountV2.ts [hunk 2]
```diff
@@ -346,6 +346,81 @@ export class BiconomySmartAccountV2 extends BaseSmartContractAccount {
...
+      verificationGasLimit,
+      maxFeePerGas
+    } = await this.buildUserOp(transactions, buildUseropDto)
+
+    const _callGasLimit = BigInt(callGasLimit || 0)
+    const _preVerificationGas = BigInt(preVerificationGas || 0)
+    const _verificationGasLimit = BigInt(verificationGasLimit || 0)
+    const _maxFeePerGas = BigInt(maxFeePerGas || 0)
+
+    if (buildUseropDto?.paymasterServiceData?.mode) {
```
Comment (VGabriel45) [inline src/account/BiconomySmartAccountV2.ts:409]: I think the conditions are reversed, if there is a paymaster it should be "(callGasLimit + 3 * verificationGasLimit + preVerificationGas) * userOp.maxFeePerGas".

If there is no paymaster it should be "(callGasLimit + preVerificationGas + verificationGasLimit) * userOp.maxFeePerGas"

File: src/account/BiconomySmartAccountV2.ts [hunk 3]
```diff
@@ -346,6 +346,81 @@ export class BiconomySmartAccountV2 extends BaseSmartContractAccount {
...
+      return (
+        _callGasLimit +
+        _preVerificationGas +
+        _verificationGasLimit * _maxFeePerGas
```
Comment (VGabriel45) [inline src/account/BiconomySmartAccountV2.ts:413]: This should be "(callGasLimit + preVerificationGas + verificationGasLimit) * userOp.maxFeePerGas", you are multiplying _verificationGasLimit with _maxFeePerGas instead.

File: src/account/BiconomySmartAccountV2.ts (hunk 2 above)
Comment (joepegler) [reply src/account/BiconomySmartAccountV2.ts]: Nice catch

**Pull request merged**